c7n/resources/appelb.py

```diff
@@ -153,33 +140,119 @@ def process_alb(self, alb):
 
 
 class AppELBListenerFilter(Filter):
-    """ Base class for filters that query listeners.
+    """ Base class for filters that query LB listeners.
     """
-    def describe_appelb_listeners(self, albs):
+    def initialize(self, albs):
         def _process_listeners(alb):
-            client = local_session(self.manager.session_factory).client('elbv2')
-            results = client.describe_listeners(
-                LoadBalancerArn=alb['LoadBalancerArn'])
-            alb['Listeners'] = results['Listeners']
+            if 'Listeners' not in alb:
+                client = local_session(
+                    self.manager.session_factory).client('elbv2')
+                results = client.describe_listeners(
+                    LoadBalancerArn=alb['LoadBalancerArn'])
+                alb['Listeners'] = results['Listeners']
 
         with self.manager.executor_factory(max_workers=2) as w:
             list(w.map(_process_listeners, albs))
 
 
-@filters.register('is-ssl')
-class AppELBIsSSLFilter(AppELBListenerFilter):
+class AppELBAttributeFilter(Filter):
+    """ Base class for filters that query LB attributes.
+    """
+    def initialize(self, albs):
+        def _process_attributes(alb):
+            if 'Attributes' not in alb:
+                client = local_session(
+                    self.manager.session_factory).client('elbv2')
+                results = client.describe_load_balancer_attributes(
+                    LoadBalancerArn=alb['LoadBalancerArn'])
+                alb['Attributes'] = results['Attributes']
+
+        with self.manager.executor_factory(max_workers=2) as w:
+            list(w.map(_process_attributes, albs))
+
+
+class AppELBTargetGroupFilter(Filter):
+    """ Base class for filters that query LB target groups.
+    """
+    def initialize(self, albs):
+        def _process_target_groups(alb):
+            if 'TargetGroups' not in alb:
+                client = local_session(
+                    self.manager.session_factory).client('elbv2')
+                results = client.describe_target_groups(
+                    LoadBalancerArn=alb['LoadBalancerArn'])
+                alb['TargetGroups'] = results['TargetGroups']
+
+                for target_group in alb['TargetGroups']:
+                    result = client.describe_target_health(
+                        TargetGroupArn=target_group['TargetGroupArn'])
+                    target_group['TargetHealthDescriptions'] = result[
+                        'TargetHealthDescriptions']
+
+        with self.manager.executor_factory(max_workers=2) as w:
+            list(w.map(_process_target_groups, albs))
+
+
+@filters.register('is-https')
+class AppELBIsHTTPSFilter(AppELBListenerFilter):
```

Comment:
rather than a separate filter per listener attribute, just making the listener and target filter bases derive from value  filter makes it more general purpose to handle this and other use cases, thoughts?